### ISSUE TITLE
subdict_match._dict_match() called thousands of times per reaction

### ISSUE THREAD

**Issue by spyd3r** [saltstack/salt#46798]
### Description of Issue/Question
When specifying tgt_type: 'pillar' in a reactor file, the function subdict_match._dict_match() gets called ~2700 times per reaction. Running salt-master in debug mode yields thousands of the following message:
Attempting to match 'role:cfg' in 'minion' using delimiter ':'
2018-03-27 16:16:58,661 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'role' using delimiter ':'
2018-03-27 16:16:58,662 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'minion:role' using delimiter ':'
2018-03-27 16:16:58,663 [salt.utils       ][DEBUG   ][1719] Attempting to match 'role:cfg' in 'minion' using delimiter ':'
2018-03-27 16:16:58,664 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'role' using delimiter ':'
2018-03-27 16:16:58,664 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'minion:role' using delimiter ':'
2018-03-27 16:16:58,666 [salt.utils       ][DEBUG   ][1719] Attempting to match 'role:cfg' in 'minion' using delimiter ':'
2018-03-27 16:16:58,666 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'role' using delimiter ':'
2018-03-27 16:16:58,667 [salt.utils       ][DEBUG   ][1719] Attempting to match 'cfg' in 'minion:role' using delimiter ':'
### Setup

Add reactor sls with a tgt of a pillar path and tgt_type of pillar

Here's one of ours:
gitlab_runner_decom:
  cmd.gitlab.delete_runner:
    - tgt: 'minion:role:cfg'
    - tgt_type: 'pillar'
    - arg:
      - {{ hostname }}

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report

Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.6 (default, Nov 23 2017, 15:49:48)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
         locale: UTF-8
        machine: x86_64
        release: 3.13.0-143-generic
         system: Linux
        version: Ubuntu 14.04 trusty

**Comment by garethgreenaway** [comment]
@spyd3r Thanks for the report.  To clarify, is the issue that you're reporting regarding the amount of logging in debug mode or is there a problem with the matching?

**Comment by spyd3r** [comment]
@garethgreenaway The issue I'm reporting is the amount of work being performed while doing pillar target matching, causing unnecessary system load on the salt master. Debug logging revealed this large amount of activity occurring. I came across this issue while debugging why my reactors don't appear to be executing with %100 consistency. For instance, I have a reactor that upon instance termination will remove the minion from server monitoring, delete its DNS entry, delete its CI/CD agent, and remove its salt minion key. Sometimes, one or more of the reactor states won't execute. I suspect that the salt master is under too much load and some jobs are being lost. I put the master into debug mode to see if I could determine what was happening with the sporadic reactor behavior. That's when I noticed the reported issue.
